src/encoding.ts

```diff
@@ -13,17 +13,31 @@
               / "*" / "+" / "," / ";" / "="
  */
 
-const excludeSubDelimiters = /[^!$'()*+,;|:]/g
-
 export type URLParamsEncodingType =
   | 'default'
   | 'uri'
   | 'uriComponent'
   | 'none'
   | 'legacy'
 
-export const encodeURIComponentExcludingSubDelims = (segment: string): string =>
-  segment.replace(excludeSubDelimiters, match => encodeURIComponent(match))
+export const encodeURIComponentExcludingSubDelims = (
+  segment: string
+): string => {
+  // Define sub-delimiters to exclude from encoding
+  const subDelimiters = /[!$'()*+,;|:]/g
+
+  // Use Array.from to correctly handle characters represented by surrogate pairs
+  return Array.from(segment)
+    .map(char => {
+      // Check if the character is a sub-delimiter
+      if (subDelimiters.test(char)) {
+        return char // Return the character as is if it's a sub-delimiter
```

Comment:
This could probably be optimized a bit more, perhaps checking for character codes instead of matching against a regex. But maybe JS engines are so smart these days that it doesn't matter. Did you ever benchmark `path-parser`?